zrml/prediction-markets/src/lib.rs

```diff
@@ -1349,6 +1350,338 @@ mod pallet {
+            let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+            Self::ensure_market_is_active(&market)?;
+            let mut early_close = market.early_close.ok_or(Error::<T>::NoEarlyCloseScheduled)?;
+            let weight = match early_close.state {
+                // market period got already reset inside `dispute_early_close`
+                EarlyCloseState::Disputed => T::WeightInfo::reject_early_close_after_dispute(),
+                EarlyCloseState::ScheduledAsOther => {
+                    // ensure we don't reject if the old market period is already over
+                    // so that we don't switch back to an invalid market period
+                    debug_assert!(
+                        {
+                            match early_close.old {
+                                MarketPeriod::Block(ref range) => {
+                                    let now_block = <frame_system::Pallet<T>>::block_number();
+                                    now_block < range.end
+                                }
+                                MarketPeriod::Timestamp(ref range) => {
+                                    let now_time = <zrml_market_commons::Pallet<T>>::now();
+                                    now_time < range.end
+                                }
+                            }
+                        },
+                        "This will never happen, because schedule_early_close ensures that the \
+                         new end is always before the old end, 
+                    otherwise the switch to an old market period would lead to a never ending \
+                         market!"
+                    );
+
+                    let ids_len_0 = Self::clear_auto_close(&market_id)?;
+
+                    <zrml_market_commons::Pallet<T>>::mutate_market(&market_id, |market| {
+                        market.period = early_close.old.clone();
+                        Ok(())
+                    })?;
+
+                    // important to do this after the market period is mutated
+                    let ids_len_1 = Self::insert_auto_close(&market_id)?;
+
+                    T::WeightInfo::reject_early_close_after_authority(ids_len_0, ids_len_1)
+                }
+                EarlyCloseState::ScheduledAsMarketCreator | EarlyCloseState::Rejected => {
+                    return Err(Error::<T>::InvalidEarlyCloseState.into());
+                }
+            };
+
+            if let Some(disputor_bond) = market.bonds.close_dispute.as_ref() {
+                let close_disputor = &disputor_bond.who;
+                Self::repatriate_close_request_bond(&market_id, close_disputor)?;
+                Self::unreserve_close_dispute_bond(&market_id)?;
+            }
+
+            <zrml_market_commons::Pallet<T>>::mutate_market(&market_id, |market| {
+                early_close.state = EarlyCloseState::Rejected;
+                market.early_close = Some(early_close);
```

Comment:
Can it happen that the market resolves at the original (first) close time, but the storage still contains `market.early_close = Some(early_close)`? In that case the final market storage element would represent an untruthful state.

Reply:
I see your point here. `early_close` of market always contains the `state` and the `old` and `new` market period. This means, if an early close request was rejected by the AC, the `market` still contains the information that an early close was rejected. If it was rejected, it does not mean that the market closed at the `new` end. The market always closes according to the `market.period` field. In addition, at the point of rejection, the market creator can't close the market early anymore. This ability belongs only to the advisory committee at that point. If we write `None` for the early close field, the market creator could schedule an early close again. Generally we had discussions about if the market creator should be able to schedule an early close again after one rejection. I don't think this is necessary and would actually put more strain on the AC, even if the market creator pays a bond each time. Just let the AC close the market early, if even required.